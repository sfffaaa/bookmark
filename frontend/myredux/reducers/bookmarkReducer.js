import {
    CREATE_BOOKMARK_ACTION_TYPE,
    LIST_BOOKMARK_ACTION_TYPE,
    DELETE_BOOKMARK_ACTION_TYPE,
    UPGRADE_BOOKMARK_ACTION_TYPE,
} from '../constants/action-types';

// [TODO] Think about mask
const bookmarkInitState = {
    data: [],
    needUpdate: false,
    errMsg: null,
};

/* eslint-disable-next-line import/prefer-default-export */
export const bookmarkReducer = (bookmarks = bookmarkInitState, action) => {
    switch (action.type) {
    case CREATE_BOOKMARK_ACTION_TYPE.PENDING:
        console.log(action);
        return bookmarks;
    case CREATE_BOOKMARK_ACTION_TYPE.SUCCESS:
        console.log(action);
        // [TODO] Add need update exist.
        return Object.assign({}, bookmarks, {
            needUpdate: true,
            errMsg: null,
        });

    case LIST_BOOKMARK_ACTION_TYPE.PENDING:
        console.log(action);
        return bookmarks;
    case LIST_BOOKMARK_ACTION_TYPE.SUCCESS:
        console.log(action);
        return Object.assign({}, bookmarks, {
            data: action.response.data.data,
            needUpdate: false,
            errMsg: null,
        });

    case DELETE_BOOKMARK_ACTION_TYPE.PENDING:
        console.log(action);
        return bookmarks;
    case DELETE_BOOKMARK_ACTION_TYPE.SUCCESS:
        console.log(action);
        return Object.assign({}, bookmarks, {
            needUpdate: true,
            errMsg: null,
        });

    case UPGRADE_BOOKMARK_ACTION_TYPE.PENDING:
        console.log(action);
        return bookmarks;
    case UPGRADE_BOOKMARK_ACTION_TYPE.SUCCESS:
        console.log(action);
        return Object.assign({}, bookmarks, {
            needUpdate: true,
            errMsg: null,
        });

    case CREATE_BOOKMARK_ACTION_TYPE.ERROR:
    case LIST_BOOKMARK_ACTION_TYPE.ERROR:
    case DELETE_BOOKMARK_ACTION_TYPE.ERROR:
    case UPGRADE_BOOKMARK_ACTION_TYPE.ERROR:
        console.log(action);
        return Object.assign({}, bookmarks, {
            needUpdate: false,
            errMsg: (action.response && action.response.errMsg)
                    || (action.error && action.error.errMsg)
                    || 'Unknown error',
        });


    default:
        return bookmarks;
    }
};
