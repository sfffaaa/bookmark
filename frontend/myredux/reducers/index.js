import { combineReducers } from 'redux';
import { bookmarkReducer } from './bookmarkReducer';

/* eslint-disable-next-line import/prefer-default-export */
export const myCombineReducers = combineReducers({
    bookmarks: bookmarkReducer,
});
