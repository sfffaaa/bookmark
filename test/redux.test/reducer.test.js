/* global describe it expect */
import deepFreeze from 'deep-freeze';

import { bookmarkReducer } from '../../frontend/myredux/reducers/bookmarkReducer';
import {
    CREATE_BOOKMARK_ACTION_TYPE,
    LIST_BOOKMARK_ACTION_TYPE,
    DELETE_BOOKMARK_ACTION_TYPE,
    UPGRADE_BOOKMARK_ACTION_TYPE,
} from '../../frontend/myredux/constants/action-types';

describe('bookmark reducer test', () => {
    it('test error occurs', () => {
        const bookmarkInitStates = [{
            data: [],
            needUpdate: false,
            errMsg: null,
        }, {
            data: ['dummy1'],
            needUpdate: true,
            errMsg: 'ewrer',
        }];

        const errorTypes = [
            CREATE_BOOKMARK_ACTION_TYPE.ERROR,
            LIST_BOOKMARK_ACTION_TYPE.ERROR,
            DELETE_BOOKMARK_ACTION_TYPE.ERROR,
            UPGRADE_BOOKMARK_ACTION_TYPE.ERROR,
        ];
        for (let j = 0; j < bookmarkInitStates.length; j += 1) {
            const bookmarkInitState = bookmarkInitStates[j];
            for (let i = 0; i < errorTypes.length; i += 1) {
                const errorType = errorTypes[i];
                expect(bookmarkReducer(
                    deepFreeze(bookmarkInitState), {
                        type: errorType,
                        response: {
                            errMsg: 'error',
                        },
                    },
                )).toMatchSnapshot();

                expect(bookmarkReducer(
                    deepFreeze(bookmarkInitState), {
                        type: errorType,
                        error: {
                            errMsg: 'error',
                        },
                    },
                )).toMatchSnapshot();

                expect(bookmarkReducer(
                    deepFreeze(bookmarkInitState), {
                        type: errorType,
                    },
                )).toMatchSnapshot();
            }
        }
    });

    it('test pending occurs', () => {
        const bookmarkInitStates = [{
            data: [],
            needUpdate: false,
            errMsg: null,
        }, {
            data: ['dummy1'],
            needUpdate: true,
            errMsg: 'ewrer',
        }];

        const pendingTypes = [
            CREATE_BOOKMARK_ACTION_TYPE.PENDING,
            LIST_BOOKMARK_ACTION_TYPE.PENDING,
            DELETE_BOOKMARK_ACTION_TYPE.PENDING,
            UPGRADE_BOOKMARK_ACTION_TYPE.PENDING,
        ];
        for (let j = 0; j < bookmarkInitStates.length; j += 1) {
            const bookmarkInitState = bookmarkInitStates[j];
            for (let i = 0; i < pendingTypes.length; i += 1) {
                const pendingType = pendingTypes[i];
                expect(bookmarkReducer(
                    deepFreeze(bookmarkInitState), {
                        type: pendingType,
                    },
                )).toMatchSnapshot();
            }
        }
    });

    it('test success occurs without list', () => {
        const bookmarkInitStates = [{
            data: ['dummy1'],
            needUpdate: false,
            errMsg: 'ewrer',
        }];
        const successTypes = [
            CREATE_BOOKMARK_ACTION_TYPE.SUCCESS,
            DELETE_BOOKMARK_ACTION_TYPE.SUCCESS,
            UPGRADE_BOOKMARK_ACTION_TYPE.SUCCESS,
        ];
        for (let j = 0; j < bookmarkInitStates.length; j += 1) {
            const bookmarkInitState = bookmarkInitStates[j];
            for (let i = 0; i < successTypes.length; i += 1) {
                const successType = successTypes[i];
                expect(bookmarkReducer(
                    deepFreeze(bookmarkInitState), {
                        type: successType,
                    },
                )).toMatchSnapshot();
            }
        }
    });

    it('test success occurs with list', () => {
        const bookmarkInitStates = [{
            data: ['dummy1'],
            needUpdate: true,
            errMsg: 'ewrer',
        }];
        for (let j = 0; j < bookmarkInitStates.length; j += 1) {
            const bookmarkInitState = bookmarkInitStates[j];
            expect(bookmarkReducer(
                deepFreeze(bookmarkInitState), {
                    type: LIST_BOOKMARK_ACTION_TYPE.SUCCESS,
                    response: { data: { data: ['1', '2', '3'] } },
                },
            )).toMatchSnapshot();
        }
    });

    it('test no action', () => {
        const bookmarkInitState = [{
            data: ['dummy1'],
            needUpdate: true,
            errMsg: 'ewrer',
        }];
        expect(bookmarkReducer(
            deepFreeze(bookmarkInitState), {
                type: '__WHAT_IS_THIS__',
                response: { data: { data: ['1', '2', '3'] } },
            },
        )).toMatchSnapshot();

        expect(bookmarkReducer(undefined, { type: '__WHAT_IS_THIS__' })).toMatchSnapshot();
    });
});
