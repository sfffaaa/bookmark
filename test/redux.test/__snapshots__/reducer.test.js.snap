// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`bookmark reducer test test error occurs 1`] = `
Object {
  "data": Array [],
  "errMsg": "error",
  "needUpdate": false,
}
`;

exports[`bookmark reducer test test error occurs 2`] = `
Object {
  "data": Array [],
  "errMsg": "error",
  "needUpdate": false,
}
`;

exports[`bookmark reducer test test error occurs 3`] = `
Object {
  "data": Array [],
  "errMsg": "Unknown error",
  "needUpdate": false,
}
`;

exports[`bookmark reducer test test error occurs 4`] = `
Object {
  "data": Array [],
  "errMsg": "error",
  "needUpdate": false,
}
`;

exports[`bookmark reducer test test error occurs 5`] = `
Object {
  "data": Array [],
  "errMsg": "error",
  "needUpdate": false,
}
`;

exports[`bookmark reducer test test error occurs 6`] = `
Object {
  "data": Array [],
  "errMsg": "Unknown error",
  "needUpdate": false,
}
`;

exports[`bookmark reducer test test error occurs 7`] = `
Object {
  "data": Array [],
  "errMsg": "error",
  "needUpdate": false,
}
`;

exports[`bookmark reducer test test error occurs 8`] = `
Object {
  "data": Array [],
  "errMsg": "error",
  "needUpdate": false,
}
`;

exports[`bookmark reducer test test error occurs 9`] = `
Object {
  "data": Array [],
  "errMsg": "Unknown error",
  "needUpdate": false,
}
`;

exports[`bookmark reducer test test error occurs 10`] = `
Object {
  "data": Array [],
  "errMsg": "error",
  "needUpdate": false,
}
`;

exports[`bookmark reducer test test error occurs 11`] = `
Object {
  "data": Array [],
  "errMsg": "error",
  "needUpdate": false,
}
`;

exports[`bookmark reducer test test error occurs 12`] = `
Object {
  "data": Array [],
  "errMsg": "Unknown error",
  "needUpdate": false,
}
`;

exports[`bookmark reducer test test error occurs 13`] = `
Object {
  "data": Array [
    "dummy1",
  ],
  "errMsg": "error",
  "needUpdate": false,
}
`;

exports[`bookmark reducer test test error occurs 14`] = `
Object {
  "data": Array [
    "dummy1",
  ],
  "errMsg": "error",
  "needUpdate": false,
}
`;

exports[`bookmark reducer test test error occurs 15`] = `
Object {
  "data": Array [
    "dummy1",
  ],
  "errMsg": "Unknown error",
  "needUpdate": false,
}
`;

exports[`bookmark reducer test test error occurs 16`] = `
Object {
  "data": Array [
    "dummy1",
  ],
  "errMsg": "error",
  "needUpdate": false,
}
`;

exports[`bookmark reducer test test error occurs 17`] = `
Object {
  "data": Array [
    "dummy1",
  ],
  "errMsg": "error",
  "needUpdate": false,
}
`;

exports[`bookmark reducer test test error occurs 18`] = `
Object {
  "data": Array [
    "dummy1",
  ],
  "errMsg": "Unknown error",
  "needUpdate": false,
}
`;

exports[`bookmark reducer test test error occurs 19`] = `
Object {
  "data": Array [
    "dummy1",
  ],
  "errMsg": "error",
  "needUpdate": false,
}
`;

exports[`bookmark reducer test test error occurs 20`] = `
Object {
  "data": Array [
    "dummy1",
  ],
  "errMsg": "error",
  "needUpdate": false,
}
`;

exports[`bookmark reducer test test error occurs 21`] = `
Object {
  "data": Array [
    "dummy1",
  ],
  "errMsg": "Unknown error",
  "needUpdate": false,
}
`;

exports[`bookmark reducer test test error occurs 22`] = `
Object {
  "data": Array [
    "dummy1",
  ],
  "errMsg": "error",
  "needUpdate": false,
}
`;

exports[`bookmark reducer test test error occurs 23`] = `
Object {
  "data": Array [
    "dummy1",
  ],
  "errMsg": "error",
  "needUpdate": false,
}
`;

exports[`bookmark reducer test test error occurs 24`] = `
Object {
  "data": Array [
    "dummy1",
  ],
  "errMsg": "Unknown error",
  "needUpdate": false,
}
`;

exports[`bookmark reducer test test no action 1`] = `
Array [
  Object {
    "data": Array [
      "dummy1",
    ],
    "errMsg": "ewrer",
    "needUpdate": true,
  },
]
`;

exports[`bookmark reducer test test no action 2`] = `
Object {
  "data": Array [],
  "errMsg": null,
  "needUpdate": false,
}
`;

exports[`bookmark reducer test test pending occurs 1`] = `
Object {
  "data": Array [],
  "errMsg": null,
  "needUpdate": false,
}
`;

exports[`bookmark reducer test test pending occurs 2`] = `
Object {
  "data": Array [],
  "errMsg": null,
  "needUpdate": false,
}
`;

exports[`bookmark reducer test test pending occurs 3`] = `
Object {
  "data": Array [],
  "errMsg": null,
  "needUpdate": false,
}
`;

exports[`bookmark reducer test test pending occurs 4`] = `
Object {
  "data": Array [],
  "errMsg": null,
  "needUpdate": false,
}
`;

exports[`bookmark reducer test test pending occurs 5`] = `
Object {
  "data": Array [
    "dummy1",
  ],
  "errMsg": "ewrer",
  "needUpdate": true,
}
`;

exports[`bookmark reducer test test pending occurs 6`] = `
Object {
  "data": Array [
    "dummy1",
  ],
  "errMsg": "ewrer",
  "needUpdate": true,
}
`;

exports[`bookmark reducer test test pending occurs 7`] = `
Object {
  "data": Array [
    "dummy1",
  ],
  "errMsg": "ewrer",
  "needUpdate": true,
}
`;

exports[`bookmark reducer test test pending occurs 8`] = `
Object {
  "data": Array [
    "dummy1",
  ],
  "errMsg": "ewrer",
  "needUpdate": true,
}
`;

exports[`bookmark reducer test test success occurs with list 1`] = `
Object {
  "data": Array [
    "1",
    "2",
    "3",
  ],
  "errMsg": null,
  "needUpdate": false,
}
`;

exports[`bookmark reducer test test success occurs without list 1`] = `
Object {
  "data": Array [
    "dummy1",
  ],
  "errMsg": null,
  "needUpdate": true,
}
`;

exports[`bookmark reducer test test success occurs without list 2`] = `
Object {
  "data": Array [
    "dummy1",
  ],
  "errMsg": null,
  "needUpdate": true,
}
`;

exports[`bookmark reducer test test success occurs without list 3`] = `
Object {
  "data": Array [
    "dummy1",
  ],
  "errMsg": null,
  "needUpdate": true,
}
`;
